File: src/value/composite.zig
//! 复合值类型模块
//!
//! 定义 Glue 语言中的复合数据结构：
//! - ArrayValue 数组（header + elements 分离分配，支持动态扩容）
//! - RecordValue 记录（连续内存：header + fields）
//! - AdtValue 代数数据类型（连续内存：header + fields）
//! - NewtypeValue 新类型包装（固定大小）
//! - Cell 可变单元格（固定大小）
//! - Range 整数区间（固定大小）
//!
//! 所有堆对象以 ObjHeader 作为首字段，统一类型识别与引用计数。

const std = @import("std");
const obj_header = @import("obj_header.zig");
const ObjHeader = obj_header.ObjHeader;
const ThreadContext = obj_header.ThreadContext;
const value = @import("mod.zig");
const Value = value.Value;

/// ADT 字段，可带名称
pub const AdtField = struct {
    name: ?[]const u8,
    value: Value,
};

/// 代数数据类型值，由类型名、构造器和字段列表组成
///
/// 连续内存布局：[AdtValue header | AdtField fields...]
/// fields 切片指向尾部连续区域，单次分配单次释放
pub const AdtValue = struct {
    header: ObjHeader = .{ .type_tag = .adt },
    type_name: []const u8,
    constructor: []const u8,
    fields: []AdtField,
    /// 字段引用标记位图：第 i 位为 1 表示第 i 个字段类型为 &T / *T。
    /// 当前限制：最多 64 个字段；超出 64 的字段保守按值语义处理。
    field_ref_bits: u64 = 0,

    /// 释放所有字段值
    /// fields 是连续内存的一部分，由 adtDeinit 中的 freeObj 统一释放
    pub fn deinit(self: *AdtValue, tctx: *ThreadContext) void {
        if (!obj_header.shutdown_mode) {
            for (self.fields) |*f| f.value.release(tctx);
        }
    }

    /// 判断指定字段是否为引用类型（&T / *T）
    pub inline fn fieldIsRef(self: *const AdtValue, field_id: u16) bool {
        if (field_id >= 64) return false;
        return (self.field_ref_bits & (@as(u64, 1) << @intCast(field_id))) != 0;
    }
};

/// 新类型值，包装一个内部值并附加类型名
pub const NewtypeValue = struct {
    header: ObjHeader = .{ .type_tag = .newtype },
    type_name: []const u8,
    inner: Value,

    /// 释放内部值
    pub fn deinit(self: *NewtypeValue, tctx: *ThreadContext) void {
        if (!obj_header.shutdown_mode) {
            self.inner.release(tctx);
        }
    }
};

/// 数组值，支持定长和变长
///
/// header 和 elements 分离分配：header 固定大小（页池），elements 可独立扩容
/// 这是唯一不使用连续内存的变长类型，因为 array_push/pop 需要替换 elements 切片
/// 同时保持 header 地址不变（tracked_objs 引用 header 指针）
pub const ArrayValue = struct {
    header: ObjHeader = .{ .type_tag = .array },
    elements: []Value,
    capacity: usize = 0,
    fixed_size: ?u64 = null,
    /// 元素类型是否为 &T / *T 引用类型。
    /// true 时元素保持引用语义（retain），false 时按值语义深拷贝。
    elem_is_ref: bool = false,

    /// 释放所有元素并回收元素数组内存
    /// header 由 arrayDeinit 中的 freeObj 释放
    /// arena 分配的对象：elements 也从 arena 分配，跳过 freeObj（arena.reset 统一回收）
    pub fn deinit(self: *ArrayValue, tctx: *ThreadContext) void {
        if (!obj_header.shutdown_mode) {
            for (self.elements) |*e| e.release(tctx);
        }
        if (self.elements.len > 0 and !self.header.isArenaAllocated()) {
            tctx.freeObj(@ptrCast(self.elements.ptr));
        }
    }
};

/// 记录值，类型名加具名字段映射
///
/// 连续内存布局：[RecordValue header | Value fields...]
/// fields 切片指向尾部连续区域，单次分配单次释放
/// 字段值可通过 set 替换（in-place 写入连续区域中的槽位）
///
/// 字段存储改为连续 `[]Value` 切片，field_id 即切片索引：
/// - ADT/newtype/error_newtype：field_id=0 是 `__tag`，1..N 是构造器字段
/// - Record literal：field_id=0..N-1 按声明顺序
/// 字段名 → field_id 的映射在 IR builder 阶段完成，运行时零字符串比较。
pub const RecordValue = struct {
    header: ObjHeader = .{ .type_tag = .record },
    type_name: []const u8,
    /// 字段值连续存储，field_id 即索引
    fields: []Value,
    /// 字段名表（与 fields 同长，仅 format/调试用；运行时热路径不访问）
    /// 索引与 fields 对齐：field_names[i] 对应 fields[i] 的名字
    field_names: []const ?[]const u8 = &.{},
    /// 字段引用标记位图：第 i 位为 1 表示第 i 个字段类型为 &T / *T。
    /// 当前限制：最多 64 个字段；超出 64 的字段保守按值语义处理。
    field_ref_bits: u64 = 0,

    /// 释放字段值
    /// fields 是连续内存的一部分，由 recordDeinit 中的 freeObj 统一释放
    /// field_names 指向源码 arena 或 string_pool，无需释放
    pub fn deinit(self: *RecordValue, tctx: *ThreadContext) void {
        if (!obj_header.shutdown_mode) {
            for (self.fields) |*f| f.release(tctx);
        }
    }

    /// 判断指定字段是否为引用类型（&T / *T）
    pub inline fn fieldIsRef(self: *const RecordValue, field_id: u16) bool {
        if (field_id >= 64) return false;
        return (self.field_ref_bits & (@as(u64, 1) << @intCast(field_id))) != 0;
    }

    /// 按 field_id 读取字段值（运行时热路径）
    pub inline fn get(self: *const RecordValue, field_id: u16) ?Value {
        if (field_id >= self.fields.len) return null;
        return self.fields[field_id];
    }

    /// 按 field_id 写入字段值（运行时热路径）
    /// 替换语义：release 旧值，retain 新值由调用方负责
    pub inline fn set(self: *RecordValue, field_id: u16, v: Value, tctx: *ThreadContext) bool {
        if (field_id >= self.fields.len) return false;
        const slot = &self.fields[field_id];
        slot.release(tctx);
        slot.* = v;
        return true;
    }
};

/// 可变单元格，提供对内部值的可变引用语义
pub const Cell = struct {
    header: ObjHeader = .{ .type_tag = .cell },
    inner: Value,

    /// 释放内部值
    pub fn deinit(self: *Cell, tctx: *ThreadContext) void {
        if (!obj_header.shutdown_mode) {
            self.inner.release(tctx);
        }
    }
};

/// 整数区间，支持半开和闭区间
pub const Range = struct {
    header: ObjHeader = .{ .type_tag = .range },
    start: [16]u8,
    end: [16]u8,
    inclusive: bool,
    start_i64: ?i64 = null,
    end_i64: ?i64 = null,

    /// 空实现：整数值为内联类型，区间本身无需释放
    pub fn deinit(self: *Range, tctx: *ThreadContext) void {
        _ = self;
        _ = tctx;
    }
};

// ── deinit_table 注册函数 ──
// 所有 deinit 包装函数：执行 Type.deinit（释放内部 RC 子对象/独立缓冲区），
// 若对象非 arena 分配则 freeObj 释放对象本体；arena 分配的对象由 arena.reset 统一回收。

pub fn adtDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *AdtValue = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    if (!obj.isArenaAllocated()) tctx.freeObj(@ptrCast(self));
}

pub fn newtypeDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *NewtypeValue = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    if (!obj.isArenaAllocated()) tctx.freeObj(@ptrCast(self));
}

pub fn arrayDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *ArrayValue = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    if (!obj.isArenaAllocated()) tctx.freeObj(@ptrCast(self));
}

pub fn recordDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *RecordValue = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    if (!obj.isArenaAllocated()) tctx.freeObj(@ptrCast(self));
}

pub fn cellDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *Cell = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    if (!obj.isArenaAllocated()) tctx.freeObj(@ptrCast(self));
}

pub fn rangeDeinit(obj: *ObjHeader, tctx: *ThreadContext) void {
    const self: *Range = @alignCast(@fieldParentPtr("header", obj));
    self.deinit(tctx);
    if (!obj.isArenaAllocated()) tctx.freeObj(@ptrCast(self));
}

/// 注册所有复合类型的 deinit 函数
pub fn registerDeinits() void {
    obj_header.registerDeinit(.adt, adtDeinit);
    obj_header.registerDeinit(.newtype, newtypeDeinit);
    obj_header.registerDeinit(.array, arrayDeinit);
    obj_header.registerDeinit(.record, recordDeinit);
    obj_header.registerDeinit(.cell, cellDeinit);
    obj_header.registerDeinit(.range, rangeDeinit);
}

File: rust/src/value/composite.rs
//! 复合堆对象类型：数组、记录、ADT、Newtype、Cell、Range
//!
//! 对应 Zig 实现中的复合值类型。可变容器使用 `RefCell`，不可变数据使用 `Rc`。
//! 所有引用 `Value` 的类型通过 `crate::value::Value` 前向引用，
//! 待 `mod.rs` 定义 `Value` 后即可编译。

use std::cell::RefCell;

use crate::value::Value;

// =========================================================================
// ArrayValue — 数组值
// =========================================================================

/// 数组值：元素可变（支持 push/pop），`fixed_size` 为 `Some` 时表示固定大小数组
#[derive(Debug, Clone)]
pub struct ArrayValue {
    /// 元素列表
    pub elements: Vec<Value>,
    /// 固定大小（`[T; N]` 中的 N），`None` 表示动态数组
    pub fixed_size: Option<u64>,
    /// 元素是否为引用类型（影响 GC 标记）
    pub elem_is_ref: bool,
}

impl ArrayValue {
    /// 创建动态数组
    pub fn new(elements: Vec<Value>) -> Self {
        Self {
            elements,
            fixed_size: None,
            elem_is_ref: false,
        }
    }

    /// 创建固定大小数组
    pub fn new_fixed(elements: Vec<Value>, size: u64) -> Self {
        Self {
            elements,
            fixed_size: Some(size),
            elem_is_ref: false,
        }
    }

    /// 元素数量
    pub fn len(&self) -> usize {
        self.elements.len()
    }

    /// 是否为空
    pub fn is_empty(&self) -> bool {
        self.elements.is_empty()
    }

    /// 按索引获取元素引用
    pub fn get(&self, index: usize) -> Option<&Value> {
        self.elements.get(index)
    }

    /// 追加元素
    pub fn push(&mut self, val: Value) {
        self.elements.push(val);
    }

    /// 弹出末尾元素
    pub fn pop(&mut self) -> Option<Value> {
        self.elements.pop()
    }
}

// =========================================================================
// RecordValue — 记录值
// =========================================================================

/// 记录字段：可选名称 + 值
#[derive(Debug, Clone)]
pub struct RecordField {
    pub name: Option<String>,
    pub value: Value,
}

/// 记录值：具名类型的结构化数据
#[derive(Debug, Clone)]
pub struct RecordValue {
    /// 类型名称
    pub type_name: String,
    /// 字段值列表（按定义顺序）
    pub fields: Vec<Value>,
    /// 字段名称列表（按定义顺序，`None` 表示匿名字段）
    pub field_names: Vec<Option<String>>,
    /// 字段引用位图：第 i 位为 1 表示第 i 个字段是引用类型
    pub field_ref_bits: u64,
}

impl RecordValue {
    /// 创建记录值
    pub fn new(type_name: String, fields: Vec<Value>, field_names: Vec<Option<String>>) -> Self {
        Self {
            type_name,
            fields,
            field_names,
            field_ref_bits: 0,
        }
    }

    /// 按索引获取字段值
    pub fn get_field(&self, index: usize) -> Option<&Value> {
        self.fields.get(index)
    }

    /// 按名称查找字段值
    pub fn find_field(&self, name: &str) -> Option<&Value> {
        for (i, field_name) in self.field_names.iter().enumerate() {
            if let Some(n) = field_name {
                if n == name {
                    return self.fields.get(i);
                }
            }
        }
        None
    }
}

// =========================================================================
// AdtValue — 代数数据类型值
// =========================================================================

/// ADT 字段：构造器的参数
#[derive(Debug, Clone)]
pub struct AdtField {
    pub name: Option<String>,
    pub value: Value,
}

/// ADT 值：代数数据类型实例（如 `Some(x)`、`Ok(v)` 等）
#[derive(Debug, Clone)]
pub struct AdtValue {
    /// 类型名称
    pub type_name: String,
    /// 构造器名称（如 `Some`、`Ok`）
    pub constructor: String,
    /// 构造器字段列表
    pub fields: Vec<AdtField>,
    /// 字段引用位图
    pub field_ref_bits: u64,
}

impl AdtValue {
    /// 创建 ADT 值
    pub fn new(type_name: String, constructor: String, fields: Vec<AdtField>) -> Self {
        Self {
            type_name,
            constructor,
            fields,
            field_ref_bits: 0,
        }
    }

    /// 按索引获取字段值
    pub fn get_field(&self, index: usize) -> Option<&Value> {
        self.fields.get(index).map(|f| &f.value)
    }

    /// 按名称查找字段值
    pub fn find_field(&self, name: &str) -> Option<&Value> {
        for field in &self.fields {
            if let Some(n) = &field.name {
                if n == name {
                    return Some(&field.value);
                }
            }
        }
        None
    }
}

// =========================================================================
// NewtypeValue — Newtype 值
// =========================================================================

/// Newtype 值：包装单个内部值的具名类型
#[derive(Debug, Clone)]
pub struct NewtypeValue {
    /// 类型名称
    pub type_name: String,
    /// 内部值
    pub inner: Value,
}

// =========================================================================
// Cell — 可变引用单元
// =========================================================================

/// Cell：可变引用单元，通过 `RefCell` 实现内部可变性
#[derive(Debug, Clone)]
pub struct Cell {
    pub inner: RefCell<Value>,
}

impl Cell {
    /// 创建包含指定值的 Cell
    pub fn new(val: Value) -> Self {
        Self {
            inner: RefCell::new(val),
        }
    }

    /// 获取值的借用引用
    pub fn get(&self) -> std::cell::Ref<Value> {
        self.inner.borrow()
    }

    /// 设置新值
    pub fn set(&self, val: Value) {
        *self.inner.borrow_mut() = val;
    }
}

// =========================================================================
// Range — 范围值
// =========================================================================

/// 范围值：表示整数区间 `[start, end)` 或 `[start, end]`
#[derive(Debug, Clone)]
pub struct Range {
    /// 起始值（包含）
    pub start: i64,
    /// 结束值
    pub end: i64,
    /// 是否包含结束值（`true` 为闭区间 `[start, end]`）
    pub inclusive: bool,
}

impl Range {
    /// 创建范围
    pub fn new(start: i64, end: i64, inclusive: bool) -> Self {
        Self { start, end, inclusive }
    }

    /// 判断值是否在范围内
    pub fn contains(&self, val: i64) -> bool {
        if self.inclusive {
            val >= self.start && val <= self.end
        } else {
            val >= self.start && val < self.end
        }
    }

    /// 范围内元素数量
    pub fn len(&self) -> usize {
        if self.inclusive {
            if self.end >= self.start {
                (self.end - self.start + 1) as usize
            } else {
                0
            }
        } else if self.end > self.start {
            (self.end - self.start) as usize
        } else {
            0
        }
    }

    /// 创建范围迭代器
    pub fn iter(&self) -> RangeIter {
        RangeIter {
            current: self.start,
            end: self.end,
            inclusive: self.inclusive,
        }
    }
}

/// 范围迭代器：逐个产生范围内的整数值
#[derive(Debug, Clone)]
pub struct RangeIter {
    /// 当前迭代位置
    pub current: i64,
    /// 结束值
    pub end: i64,
    /// 是否包含结束值
    pub inclusive: bool,
}
